### ISSUE TITLE
Improved syntax highlighting for inline-bold/mono/italic

### ISSUE THREAD

**Issue by rdmueller** [asciidoctor/asciidoctor-intellij-plugin#244]
Syntax highlighting for some formattings would be great. 
For example Headlines, Bold, Pre-formatted, Blocks

**Comment by ahus1** [comment]
Some highlighting is already available, this includes headings and listings. 

Headings are dark blue (in light theme), listings are shown with light gray background (when encapsulated with "----")

The first step to add formatting for bold/italic/etc. would be extending the parser in asciidoc.flex. I found Asciidoc quite hard to parse with a formal grammar. I'm happy for anyone who joins in here.

**Comment by mojavelinux** [comment]
> I found Asciidoc quite hard to parse with a formal grammar. 

Until we complete the spec, there isn't a formal grammar for AsciiDoc. However, you could reference some prior work in the following places:

* https://github.com/asciidoctor/atom-language-asciidoc/blob/master/grammars/language-asciidoc.cson
* https://github.com/Mogztter/asciidoctor-inline-parser/tree/master/lib/asciidoctor/inline_parser

**Comment by rdmueller** [comment]
@ahus1 can you point me to the code location where the current highlighting takes place?

**Comment by ahus1** [comment]
This is the Grammar: https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/master/src/main/java/org/asciidoc/intellij/lexer/asciidoc.flex

This maps attributes to highlighting: https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/master/src/main/java/org/asciidoc/intellij/lexer/AsciiDocHighlighter.java

These are all PSI Elements: https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/master/src/main/java/org/asciidoc/intellij/lexer/AsciiDocTokenTypes.java

Once the Lexer is complete, the Parser creates an PSI (I would say a AST): https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/master/src/main/java/org/asciidoc/intellij/parser/AsciiDocParserImpl.java

There are unit tests for the lexer: https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/master/src/test/java/org/asciidoc/intellij/lexer/AsciiDocLexerTest.java

There are tests for the PSI result: https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/master/src/test/java/org/asciidoc/intellij/psi/AsciiDocPsiTest.java

**Comment by ahus1** [comment]
You'll find a previous approach here, #163 - this fell short w/ multiline monospace. Maybe some re-structuring of the existing is necessary :-(

**Comment by ahus1** [comment]
I've given italic and bold highlighting a start

**Comment by ahus1** [comment]
This turned out to be less of an introduction to the parser, but now I hope this work. 

It is available from release 0.25.7 and is available from the [Github Releases tab](https://github.com/asciidoctor/asciidoctor-intellij-plugin/releases) as a download.

I re-scope this ticket to italic/bold/mono font styling. I'd like to keep all feedback for this feature in this ticket. Once this works sufficiently well, please open a new issue for the next important syntax highlighting you are missing.

Please comment in this ticket if this parsing breaks previous highlighting or misses parsing some bold/mono/italic formattings.

**Comment by rdmueller** [comment]
wow. very cool. 

it seems that it still needs some special cases:

```
bold *constrained* & **un**constrained

italic _constrained_ & __un__constrained

bold italic *_constrained_* & **__un__**constrained

monospace `constrained` & ``un``constrained

monospace bold `*constrained*` & ``**un**``constrained

monospace italic `_constrained_` & ``__un__``constrained

monospace bold italic `*_constrained_*` & ``**__un__**``constrained
```

And unordered lists which start with a `*` or `**` should also not be bold....

**Comment by ahus1** [comment]
Thanks for providing the additional test cases. A new updated version 0.25.8 is available from the [Github Releases tab](https://github.com/asciidoctor/asciidoctor-intellij-plugin/releases) as a download.

**Comment by rdmueller** [comment]
wow. that was fast!

So, it is now much better, but some edgecases seem still to be wrong:

```
monospace bold `*constrained*` & ``**un**``constrained

monospace italic `_constrained_` & ``__un__``constrained

monospace bold italic `*_constrained_*` & ``**__un__**``constrained
```

But I guess nobody uses them :-)

What is more annoying are lists... take a look at this:

```
* aasa
* basd
* csdassd
```

that is quite funny since it now switches between normal and bold when I type...

I guess you have to include all  `/^[*]+ /` from your highlighting algorithm...

**Comment by ahus1** [comment]
I polished the parsing a bit more. A new updated version 0.25.9 is available from the [Github Releases tab](https://github.com/asciidoctor/asciidoctor-intellij-plugin/releases) as a download. The cursor will also highlight matching formatter characters (like it usually does with bracket). 

As always, feedback is welcome.

**Comment by rdmueller** [comment]
eh - wow. works better than specified! :-)

I guess the styling of blocks (sidebar etc) will be now a piece of cake ;-)

**Comment by ahus1** [comment]
Thanks for the input, test cases and manual tests. The new version is now published via the JetBrains plugin repository.
